rbc/tests/test_omnisci_device_selection.py

```diff
@@ -0,0 +1,215 @@
+    if func == 'gpu':
+        return gpu_device_code if omnisci.has_cuda else err_device_code
+    if func == 'both':
+        return gpu_device_code if omnisci.has_cuda else cpu_device_code
+    raise NotImplementedError(repr((kind, func)))
+
+
+def get_pair_result(omnisci, kind1, func1, kind2, func2):
+    r1 = get_single_result(omnisci, kind1, func1)
+    r2 = get_single_result(omnisci, kind2, func2)
+    if r1 == err_device_code or r2 == err_device_code:
+        return err_device_code
+    if func1 == 'cpu' and func2 == 'gpu':
+        return err_device_code
+    if func1 == 'gpu' and func2 == 'cpu':
+        return err_device_code
+    if r1 == cpu_device_code or r2 == cpu_device_code:
+        if any_device_code not in [r1, r2]:
+            return cpu_device_code, cpu_device_code
+    return r1, r2
+
+
+def execute1(omnisci, query):
+    try:
+        _, result = omnisci.sql_execute(query)
+    except Exception as msg:
+        return msg
+    else:
+        result = list(result)
+        return decode(result[0][0])
+
+
+def execute2(omnisci, query):
+    try:
+        _, result = omnisci.sql_execute(query)
+    except Exception as msg:
+        return msg
+    else:
+        result = list(result)
+        return decode(result[0][0]), decode(result[0][1])
+
+
+def get_worker1(omnisci, ext, kind, func):
+    expected = decode(get_single_result(omnisci, kind, func))
+    if ext == 'udf':
+        query = f'select {kind}_device_selection_{ext}_{func}(0)'
+        return execute1, query, expected
+    if ext == 'udtf':
+        query = (f'select out0 from table({kind}_device_selection_{ext}_{func}'
+                 f'(cursor(select i4 from {omnisci.table_name})))')
+        return execute1, query, expected
+    raise NotImplementedError(repr((ext, kind, func)))
+
+
+def get_worker2(omnisci, ext, kind1, func1, kind2, func2):
+    expected = get_pair_result(omnisci, kind1, func1, kind2, func2)
+    if isinstance(expected, tuple):
+        expected = tuple(map(decode, expected))
+    else:
+        expected = decode(expected)
+    if ext == 'udf/udf':
+        query = (f'select {kind1}_device_selection_udf_{func1}(0),'
+                 f' {kind2}_device_selection_udf_{func2}(0)')
+        return execute2, query, expected
+    if ext == 'udtf/udf':
+        query = (f'select out0 from table({kind1}_device_selection_udtf_{func1}'
+                 f'(cursor(select {kind2}_device_selection_udf_{func2}(i4)'
+                 f' from {omnisci.table_name})))')
+        return execute1, query, expected[0] if isinstance(expected, tuple) else expected
+    raise NotImplementedError(repr((ext, kind1, func1, kind2, func2)))
+
+
+def define(omnisci):
+
+    @omnisci('int32(int32)')
+    def rt_device_selection_udf_any(x):
+        # cannot determine which device is actually used
+        return any_device_code
+
+    @omnisci('int32(int32)', devices=['cpu'])
+    def rt_device_selection_udf_cpu(x):
+        return cpu_device_code
```

Comment:
Add `devices` option to `RemoteOmnisci` class docstring.

Reply:
`devices` is actually a `Signature` argument it is documented there.

I think we need some kind of RBC/Omnisci tutorial document where we could document available features.

Reply:
Resolving this via https://github.com/xnd-project/rbc/issues/233

Reply:
Maybe add this to one of the notebooks we have